src/Simulation/Simulators/OpenSystemsSimulator/QubitManager.cs

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Linq;
+using Microsoft.Quantum.Simulation.Core;
+using Microsoft.Quantum.Simulation.Common;
+using System.Runtime.InteropServices;
+using System.Threading;
+using Microsoft.Quantum.Simulation.Simulators.Exceptions;
+using Microsoft.Quantum.Intrinsic.Interfaces;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Quantum.Experimental
+{
+    public partial class OpenSystemsSimulator
+    {
+
+        public class OpenSystemsQubitManager : QubitManager
+        {
+            private readonly OpenSystemsSimulator Parent;
+            public OpenSystemsQubitManager(OpenSystemsSimulator parent, uint capacity)
+                : base(capacity)
+            {
+                this.Parent = parent;
+            }
+
+            protected override void Release(Qubit qubit, bool wasUsedOnlyForBorrowing)
+            {
+                if (qubit != null && qubit.IsMeasured)
+                {
+                    // Try to reset measured qubits.
+                    // TODO: There are better ways to do this; increment on the
+                    //       design and make it customizable.
+                    // FIXME: In particular, this implementation uses a lot of
+                    //        extraneous measurements.
+                    if ((this.Parent as IIntrinsicMeasure).Body(new QArray<Pauli>(Pauli.PauliZ), new QArray<Qubit>(qubit)) == Result.One)
+                    {
+                        (this.Parent as IIntrinsicX).Body(qubit);
+                    }
+                }
+                base.Release(qubit, wasUsedOnlyForBorrowing);
```

Comment:
I'm a little bit confused here... for qubits that have been measured the logic above will reset them from the One state to the Zero state (or leave them in Zero if they are already there). Otherwise, the qubit goes straight to release without any further simulator calls. Doesn't this mean that a qubit that is entangled may get "leaked" such that the simlator still tracks it but the qubit manager thinks it's available for reuse?

Reply:
Right, if I understand correctly that's the correct behavior here. If the qubit has not been measured, then the developer is implicitly promising that they have coherently reset it. On a full-state simulator, we can check that that condition holds, as per your PRs to that effect, but on a noisy simulator, a qubit could validly be entangled with something as a result of noise causing a coherent unpreparation to not work correctly. In that case, we explicitly want to track and simulate the errors caused by that leakage.

Reply:
Ah, right! That does make sense, thanks for explaining.

Reply:
No worries, let me go on and add a comment to that effect... it took me a moment to remember the reasoning here, so at the least it's a good reminder to myself!